extensions/amp-embedly/0.1/amp-embedly-impl.js

```diff
@@ -0,0 +1,163 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {assertHttpsUrl} from '../../../src/url';
+import {getIframe} from '../../../src/3p-frame';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {user} from '../../../src/log';
+
+/** @const {string} */
+export const TAG = 'amp-embedly';
+
+/**
+ * Regex used to extract src from returned oEmbed data.
+ * @const {RegExp}
+ * */
+const SRC_REGEXP = /src="([^"]+)"/;
+
+/**
+ * oEmbed resource types.
+ * @const @enum {string}
+ */
+const ResourceType = {
+  LINK: 'link',
+  PHOTO: 'photo',
+  RICH: 'rich',
+  VIDEO: 'video',
+};
+
+/**
+ * Implementation of the amp-embedly component.
+ * See {@link ../amp-embedly.md} for the spec.
+ */
+export class AmpEmbedly extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?string}  */
+    this.url_ = null;
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.url_ = user().assert(
+        this.element.getAttribute('data-url'),
+        `The data-url attribute is required for <${TAG}> %s`,
+        this.element
+    );
+  }
+
+  /** @override */
+  layoutCallback() {
+    const iframe = getIframe(this.win, this.element, 'embedly');
+    this.iframe_ = iframe;
+
+    return Services.embedlyServiceForDoc(this.element)
+        .then(service => {
+          service.fetchOembedData(/**@type {string}*/(this.url_)).then(data => {
```

Comment:
This promise is detached from the chain, `layoutCallback` will not wait for `loadPromise` in this case. Either `return` or make the fat-arrow function bracket-less to return expression.